hamilton/plugins/pandas_extensions.py

```diff
@@ -229,6 +243,157 @@ def name(cls) -> str:
+    lines: bool = False
+    nrows: Optional[int] = None
+    orient: Optional[str] = None
+    precise_float: bool = False
+    storage_options: Optional[StorageOptions] = None
+    typ: Literal["frame", "series"] = "frame"
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return [DATAFRAME_TYPE]
+
+    def _get_loading_kwargs(self) -> Dict[str, Any]:
+        kwargs = {}
+        if self.chunksize is not None:
+            kwargs["chunksize"] = self.chunksize
+        if self.compression is not None:
+            kwargs["compression"] = self.compression
+        if self.convert_axes is not None:
+            kwargs["convert_axes"] = self.convert_axes
+        if self.convert_dates is not None:
+            kwargs["convert_dates"] = self.convert_dates
+        if self.date_unit is not None:
+            kwargs["date_unit"] = self.date_unit
+        if self.dtype is not None:
+            kwargs["dtype"] = self.dtype
+        if self.dtype_backend is not None:
+            kwargs["dtype_backend"] = self.dtype_backend
+        if self.encoding is not None:
+            kwargs["encoding"] = self.encoding
+        if self.encoding_errors is not None:
+            kwargs["encoding_errors"] = self.encoding_errors
+        if self.engine is not None:
+            kwargs["engine"] = self.engine
+        if self.keep_default_dates is not None:
+            kwargs["keep_default_dates"] = self.keep_default_dates
+        if self.lines is not None:
+            kwargs["lines"] = self.lines
+        if self.nrows is not None:
+            kwargs["nrows"] = self.nrows
+        if self.orient is not None:
+            kwargs["orient"] = self.orient
+        if self.precise_float is not None:
+            kwargs["precise_float"] = self.precise_float
+        if self.storage_options is not None:
+            kwargs["storage_options"] = self.storage_options
+        if self.typ is not None:
+            kwargs["typ"] = self.typ
+        return kwargs
+
+    def load_data(self, type_: Type) -> Tuple[DATAFRAME_TYPE, Dict[str, Any]]:
+        df = pd.read_json(self.filepath_or_buffer, **self._get_loading_kwargs())
+        metadata = utils.get_file_metadata(self.filepath_or_buffer)
+        return df, metadata
+
+    @classmethod
+    def name(cls) -> str:
+        return "json"
+
+
+@dataclasses.dataclass
+class PandasJsonWriter(DataSaver):
+    """Class specifically to handle saving JSON files/buffers with Pandas.
+
+    Disclaimer: We're exposing all the *current* params from the Pandas DataFrame.to_json method.
+    Some of these params may get deprecated or new params may be introduced. In the event that
+    the params/kwargs below become outdated, please raise an issue or submit a pull request.
+
+    Should map to https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_json.html
+    """
+
+    filepath_or_buffer: Optional[Union[FilePath, WriteBuffer[bytes], WriteBuffer[str]]] = None
```

Comment:
also I think this shouldn't be optional or have a default of None, right?

Reply:
otherwise in python 3.9 -- I needed to change `FilePath` to be just `str` because typing broke inspecting this class...

Reply:
> also I think this shouldn't be optional or have a default of None, right?

Good catch! I removed the default value but left the `Optional` type annotation for the reasons below. I can also remove the `Optional` type annotation if you prefer. ``

I was emulating the Pandas `to_json` type annotations found [here](https://github.com/pandas-dev/pandas/blob/ba1cccd19da778f0c3a7d6a885685da16a072870/pandas/io/json/_json.py#L136).

`path_or_buf: FilePath | WriteBuffer[str] | WriteBuffer[bytes] | None,`
`"""Optional[X] is equivalent to Union[X, None]."""`

Reply:
> > also I think this shouldn't be optional or have a default of None, right?
> 
> Good catch! I removed the default value but left the `Optional` type annotation [**1**] to emulate the Pandas `to_json` type annotation [**2**]. Though, I can remove the `Optional` type annotation if you prefer.
> 
> * [**1**] `"""Optional[X] is equivalent to Union[X, None]."""`
> * [**2**] `path_or_buf: FilePath | WriteBuffer[str] | WriteBuffer[bytes] | None,` [source](https://github.com/pandas-dev/pandas/blob/ba1cccd19da778f0c3a7d6a885685da16a072870/pandas/io/json/_json.py#L136).

Ah TIL. So yeah that's because it can return just the JSON string -- https://github.com/pandas-dev/pandas/blob/ba1cccd19da778f0c3a7d6a885685da16a072870/pandas/io/json/_json.py#L210C1-L217C17. Which I don't think makes sense here for the materializer. So let's not make it optional at all, and always required `path_or_buf`.

Reply:
Done 😄